File: backend/app/schemas/daily_care.py
"""
Daily Care Dashboard Aggregate schema — Phase P3 Daily Care Dashboard + Adherence Experience.

Architecture:
  DailyCareDashboardAggregate is a READ-ONLY aggregate that composes existing domain data.
  It does NOT create new database tables or duplicate source-of-truth data.

  Sections:
  1. MedicinesDueToday — derived from PatientMedicationSchedule + MedicationAdherenceLog
  2. UpcomingCare — derived from appointments, follow-ups, communications
  3. ContinueCare — derived from recent doctors, prescriptions
  4. HealthTimelinePreview — mini preview cards from recent encounters
  5. QuickActions — static + dynamic action shortcuts

CRITICAL:
  - Patient access is strictly scoped to their own data
  - Prescription data is NEVER mutated through this aggregate
  - Adherence actions affect tracking ONLY
  - No shadow adherence systems are created
"""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ═════════════════════════════════════════════════════════════════════════════
# SECTION 1 — Medicines Due Today
# ═════════════════════════════════════════════════════════════════════════════


class MedicationDueItem(BaseModel):
    """A single medication due for the patient today."""

    model_config = ConfigDict(from_attributes=True)

    schedule_id: UUID
    medicine_name: str
    dosage: str | None = None
    frequency: str | None = None
    instructions: str | None = None
    reminder_times: list[str] = Field(default_factory=list)

    # Adherence state for today
    adherence_status: str = "pending"  # "pending" | "taken" | "skipped" | "snoozed"
    scheduled_time: str | None = Field(
        default=None,
        description="Which reminder slot this corresponds to (HH:MM)",
    )

    # Prescription reference (immutable)
    prescription_id: UUID
    prescription_item_id: UUID


class MedicinesDueToday(BaseModel):
    """Grouped medications due today."""

    model_config = ConfigDict(from_attributes=True)

    due_now: list[MedicationDueItem] = Field(
        default_factory=list,
        description="Medications due within the current hour window",
    )
    upcoming: list[MedicationDueItem] = Field(
        default_factory=list,
        description="Medications due later today",
    )
    overdue: list[MedicationDueItem] = Field(
        default_factory=list,
        description="Medications past their scheduled time without action",
    )
    completed: list[MedicationDueItem] = Field(
        default_factory=list,
        description="Medications already taken today",
    )
    total_due: int = 0
    taken_count: int = 0
    skipped_count: int = 0
    pending_count: int = 0

    # Lightweight adherence metrics
    adherence_rate_today: float = 0.0  # 0.0 - 1.0
    current_streak_days: int = 0
    longest_streak_days: int = 0
    week_adherence_rate: float = 0.0  # 0.0 - 1.0 over last 7 days

    # TODO: Phase 4 — Apple Health / Google Fit sync status
    # health_kit_synced: bool = False
    # google_fit_synced: bool = False

    # TODO: Phase 4 — Wearable sync status
    # wearable_connected: bool = False

    # TODO: Phase 4 — Family/dependent adherence overview
    # dependent_adherence: list[DependentAdherenceSummary] = []


# ═════════════════════════════════════════════════════════════════════════════
# SECTION 2 — Upcoming Care
# ═════════════════════════════════════════════════════════════════════════════


class UpcomingAppointmentBrief(BaseModel):
    """Brief upcoming appointment for the dashboard."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    appointment_time: datetime
    doctor_name: str
    doctor_specialization: str | None = None
    clinic_name: str | None = None
    status: str


class FollowUpBrief(BaseModel):
    """Brief follow-up for the dashboard."""

    model_config = ConfigDict(from_attributes=True)

    appointment_id: UUID
    doctor_name: str
    follow_up_date: datetime
    follow_up_notes: str | None = None
    is_overdue: bool = False


class UpcomingCare(BaseModel):
    """Upcoming care section of the dashboard."""

    model_config = ConfigDict(from_attributes=True)

    next_appointment: UpcomingAppointmentBrief | None = None
    upcoming_follow_ups: list[FollowUpBrief] = Field(default_factory=list)
    overdue_follow_ups: list[FollowUpBrief] = Field(default_factory=list)
    unread_communications: int = 0
    pending_care_tasks: int = 0

    # TODO: Phase 4 — Telemedicine link for next appointment
    # telemedicine_url: str | None = None

    # TODO: Phase 4 — Pre-visit instructions
    # pre_visit_instructions: str | None = None


# ═════════════════════════════════════════════════════════════════════════════
# SECTION 3 — Continue Care
# ═════════════════════════════════════════════════════════════════════════════


class RecentDoctorBrief(BaseModel):
    """Brief doctor info for continue care."""

    model_config = ConfigDict(from_attributes=True)

    doctor_id: UUID
    doctor_name: str
    specialization: str | None = None
    clinic_name: str | None = None
    last_visit: datetime | None = None
    has_prescription: bool = False


class RecentPrescriptionBrief(BaseModel):
    """Brief prescription info for continue care."""

    model_config = ConfigDict(from_attributes=True)

    prescription_id: UUID
    appointment_id: UUID
    doctor_name: str | None = None
    created_at: datetime
    medicine_count: int = 0


class ContinueCare(BaseModel):
    """Continue care section of the dashboard."""

    model_config = ConfigDict(from_attributes=True)

    recent_doctors: list[RecentDoctorBrief] = Field(
        default_factory=list,
        description="Doctors the patient has visited recently, for rebooking",
    )
    recent_prescriptions: list[RecentPrescriptionBrief] = Field(
        default_factory=list,
        description="Recent prescriptions for download/reference",
    )

    # TODO: Phase 4 — Active treatment plans
    # active_treatment_plans: list[TreatmentPlanBrief] = []

    # TODO: Phase 4 — Chronic disease programs
    # chronic_programs: list[ChronicProgramBrief] = []


# ═════════════════════════════════════════════════════════════════════════════
# SECTION 4 — Health Timeline Preview
# ═════════════════════════════════════════════════════════════════════════════


class TimelinePreviewCard(BaseModel):
    """Mini preview card from a recent encounter."""

    model_config = ConfigDict(from_attributes=True)

    appointment_id: UUID
    appointment_time: datetime
    doctor_name: str
    doctor_specialization: str | None = None
    diagnosis: str | None = None
    treatment_summary: str | None = None
    has_prescription: bool = False
    has_encounter_summary: bool = False


class HealthTimelinePreview(BaseModel):
    """Health timeline preview section of the dashboard."""

    model_config = ConfigDict(from_attributes=True)

    recent_cards: list[TimelinePreviewCard] = Field(
        default_factory=list,
        description="Last 3 encounter preview cards",
    )
    total_encounters: int = 0

    # TODO: Phase 4 — Upcoming scheduled encounters
    # upcoming_encounters: list[TimelinePreviewCard] = []


# ═════════════════════════════════════════════════════════════════════════════
# SECTION 5 — Quick Actions
# ═════════════════════════════════════════════════════════════════════════════


class QuickAction(BaseModel):
    """A quick action shortcut for the dashboard."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    label: str
    icon: str  # lucide icon name
    route: str
    description: str | None = None
    is_future: bool = False  # True for TODO future-ready actions


class QuickActions(BaseModel):
    """Quick actions section of the dashboard."""

    model_config = ConfigDict(from_attributes=True)

    actions: list[QuickAction] = Field(default_factory=list)


# ═════════════════════════════════════════════════════════════════════════════
# DASHBOARD AGGREGATE
# ═════════════════════════════════════════════════════════════════════════════


class DailyCareDashboardAggregate(BaseModel):
    """
    Canonical READ-ONLY aggregate for the Today's Care Dashboard.

    This is the SINGLE entry point for the patient's daily care view.
    It composes existing domain data — it does NOT create new tables.

    Sections:
    1. MedicinesDueToday — adherence tracking for today's medications
    2. UpcomingCare — next appointment, follow-ups, pending tasks
    3. ContinueCare — recent doctors, prescriptions for continuity
    4. HealthTimelinePreview — mini preview cards from recent encounters
    5. QuickActions — action shortcuts

    CRITICAL:
    - Patient access is strictly scoped to their own data
    - Prescription data is NEVER mutated through this aggregate
    - Adherence actions affect tracking ONLY
    - No shadow adherence systems are created
    """

    model_config = ConfigDict(from_attributes=True)

    # ── Section 1: Medicines Due Today ─────────────────────────────────────
    medicines_due_today: MedicinesDueToday = Field(
        default_factory=MedicinesDueToday
    )

    # ── Section 2: Upcoming Care ───────────────────────────────────────────
    upcoming_care: UpcomingCare = Field(default_factory=UpcomingCare)

    # ── Section 3: Continue Care ───────────────────────────────────────────
    continue_care: ContinueCare = Field(default_factory=ContinueCare)

    # ── Section 4: Health Timeline Preview ─────────────────────────────────
    health_timeline_preview: HealthTimelinePreview = Field(
        default_factory=HealthTimelinePreview
    )

    # ── Section 5: Quick Actions ───────────────────────────────────────────
    quick_actions: QuickActions = Field(default_factory=QuickActions)

    # ══════════════════════════════════════════════════════════════════════
    # FUTURE EXTENSION HOOKS (Phase 4+)
    # ══════════════════════════════════════════════════════════════════════
    #
    # TODO: Wearable integration — daily activity summary
    # TODO: Refill reminders — prescriptions nearing depletion
    # TODO: Pharmacy ordering — one-tap refill request
    # TODO: Family medication management — caregiver view
    # TODO: Chronic disease programs — condition-specific care plans
    # TODO: Vaccination schedules — upcoming vaccine reminders
    # TODO: Apple Health / Google Fit sync — biometric data
    # TODO: Smart notification batching — grouped by urgency

File: frontend/src/services/dailyCare.ts
/**
 * Daily Care Dashboard API service — Phase P3 Daily Care Dashboard + Adherence Experience.
 *
 * Provides access to the DailyCareDashboardAggregate which composes existing
 * domain data into a patient-friendly Today's Care Dashboard.
 *
 * CRITICAL:
 * - Patient access is strictly scoped to their own data
 * - Prescription data is NEVER mutated through this aggregate
 * - Adherence actions affect tracking ONLY
 * - No shadow adherence systems are created
 */

import { api } from './api';

// ═════════════════════════════════════════════════════════════════════════════
// Types
// ═════════════════════════════════════════════════════════════════════════════

export interface MedicationDueItem {
  schedule_id: string;
  medicine_name: string;
  dosage: string | null;
  frequency: string | null;
  instructions: string | null;
  reminder_times: string[];
  adherence_status: 'pending' | 'taken' | 'skipped' | 'snoozed';
  scheduled_time: string | null;
  prescription_id: string;
  prescription_item_id: string;
}

export interface MedicinesDueToday {
  due_now: MedicationDueItem[];
  upcoming: MedicationDueItem[];
  overdue: MedicationDueItem[];
  completed: MedicationDueItem[];
  total_due: number;
  taken_count: number;
  skipped_count: number;
  pending_count: number;
  adherence_rate_today: number;
  current_streak_days: number;
  longest_streak_days: number;
  week_adherence_rate: number;
}

export interface UpcomingAppointmentBrief {
  id: string;
  appointment_time: string;
  doctor_name: string;
  doctor_specialization: string | null;
  clinic_name: string | null;
  status: string;
}

export interface FollowUpBrief {
  appointment_id: string;
  doctor_name: string;
  follow_up_date: string;
  follow_up_notes: string | null;
  is_overdue: boolean;
}

export interface UpcomingCare {
  next_appointment: UpcomingAppointmentBrief | null;
  upcoming_follow_ups: FollowUpBrief[];
  overdue_follow_ups: FollowUpBrief[];
  unread_communications: number;
  pending_care_tasks: number;
}

export interface RecentDoctorBrief {
  doctor_id: string;
  doctor_name: string;
  specialization: string | null;
  clinic_name: string | null;
  last_visit: string | null;
  has_prescription: boolean;
}

export interface RecentPrescriptionBrief {
  prescription_id: string;
  appointment_id: string;
  doctor_name: string | null;
  created_at: string;
  medicine_count: number;
}

export interface ContinueCare {
  recent_doctors: RecentDoctorBrief[];
  recent_prescriptions: RecentPrescriptionBrief[];
}

export interface TimelinePreviewCard {
  appointment_id: string;
  appointment_time: string;
  doctor_name: string;
  doctor_specialization: string | null;
  diagnosis: string | null;
  treatment_summary: string | null;
  has_prescription: boolean;
  has_encounter_summary: boolean;
}

export interface HealthTimelinePreview {
  recent_cards: TimelinePreviewCard[];
  total_encounters: number;
}

export interface QuickAction {
  id: string;
  label: string;
  icon: string;
  route: string;
  description: string | null;
  is_future: boolean;
}

export interface QuickActions {
  actions: QuickAction[];
}

export interface DailyCareDashboardAggregate {
  medicines_due_today: MedicinesDueToday;
  upcoming_care: UpcomingCare;
  continue_care: ContinueCare;
  health_timeline_preview: HealthTimelinePreview;
  quick_actions: QuickActions;
}

// ═════════════════════════════════════════════════════════════════════════════
// API calls
// ═════════════════════════════════════════════════════════════════════════════

export const dailyCareApi = {
  /**
   * Get the full Daily Care Dashboard for the authenticated patient.
   *
   * Returns a comprehensive daily care dashboard with 5 sections:
   * 1. MedicinesDueToday — medications grouped by due now / upcoming / overdue / completed
   * 2. UpcomingCare — next appointment, follow-ups, unread communications
   * 3. ContinueCare — recent doctors, recent prescriptions
   * 4. HealthTimelinePreview — last 3 encounter preview cards
   * 5. QuickActions — action shortcuts
   */
  getDashboard: async (): Promise<DailyCareDashboardAggregate> => {
    const response = await api.get('/patient/daily-care');
    return response.data;
  },
};
